File: frontend-svelte/src/lib/stores/documents.ts
/**
 * Documents store - manages document state
 */

import { writable, derived } from 'svelte/store';
import { api } from '$utils/api';

export interface Document {
	id: string;
	userId: string;
	organizationId: string;
	title: string;
	sections: Record<string, any>;
	format: string;
	domain: string | null;
	version: string;
	isActive: boolean;
	conversationId: string;
	goalTitle: string | null;
	goalId: string | null;
	name: string | null;
	cells: Record<string, any> | null;
	cascadeRules: Record<string, any> | null;
	createdAt: Date;
	lastUpdatedAt: Date;
}

interface DocumentsState {
	documents: Document[];
	currentDocument: Document | null;
	total: number;
	isLoading: boolean;
	error: string | null;
}

const initialState: DocumentsState = {
	documents: [],
	currentDocument: null,
	total: 0,
	isLoading: false,
	error: null,
};

function transformDocument(d: any): Document {
	return {
		id: d.id,
		userId: d.user_id,
		organizationId: d.organization_id,
		title: d.title,
		sections: d.sections,
		format: d.format,
		domain: d.domain,
		version: d.version,
		isActive: d.is_active,
		conversationId: d.conversation_id,
		goalTitle: d.goal_title,
		goalId: d.goal_id,
		name: d.name,
		cells: d.cells,
		cascadeRules: d.cascade_rules,
		createdAt: new Date(d.created_at),
		lastUpdatedAt: new Date(d.last_updated_at),
	};
}

function createDocumentsStore() {
	const { subscribe, set, update } = writable<DocumentsState>(initialState);

	return {
		subscribe,

		async loadDocuments(options: { limit?: number; offset?: number; domain?: string } = {}) {
			update(state => ({ ...state, isLoading: true }));
			try {
				const params = new URLSearchParams();
				if (options.limit) params.append('limit', options.limit.toString());
				if (options.offset) params.append('offset', options.offset.toString());
				if (options.domain) params.append('domain', options.domain);

				const response = await api.get<{ documents: any[]; total: number }>(
					`/api/documents/?${params.toString()}`
				);

				update(state => ({
					...state,
					documents: response.documents.map(transformDocument),
					total: response.total,
					isLoading: false,
				}));
			} catch (error: any) {
				update(state => ({
					...state,
					error: error.message,
					isLoading: false,
				}));
			}
		},

		async loadDocument(documentId: string) {
			update(state => ({ ...state, isLoading: true }));
			try {
				const response = await api.get<any>(`/api/documents/${documentId}`);
				const document = transformDocument(response);
				update(state => ({
					...state,
					currentDocument: document,
					isLoading: false,
				}));
				return document;
			} catch (error: any) {
				update(state => ({
					...state,
					error: error.message,
					isLoading: false,
				}));
				return null;
			}
		},

		async createDocument(data: {
			title: string;
			sections: Record<string, any>;
			conversationId: string;
			format?: string;
			domain?: string;
			goalTitle?: string;
			goalId?: string;
		}) {
			update(state => ({ ...state, isLoading: true }));
			try {
				const response = await api.post<any>('/api/documents/', {
					title: data.title,
					sections: data.sections,
					conversation_id: data.conversationId,
					format: data.format || 'structured_json',
					domain: data.domain,
					goal_title: data.goalTitle,
					goal_id: data.goalId,
				});
				const document = transformDocument(response);
				update(state => ({
					...state,
					documents: [document, ...state.documents],
					currentDocument: document,
					total: state.total + 1,
					isLoading: false,
				}));
				return document;
			} catch (error: any) {
				update(state => ({
					...state,
					error: error.message,
					isLoading: false,
				}));
				return null;
			}
		},

		async updateDocument(
			documentId: string,
			updates: {
				title?: string;
				sections?: Record<string, any>;
				domain?: string;
				cells?: Record<string, any>;
				cascadeRules?: Record<string, any>;
			}
		) {
			try {
				const response = await api.patch<any>(`/api/documents/${documentId}`, {
					title: updates.title,
					sections: updates.sections,
					domain: updates.domain,
					cells: updates.cells,
					cascade_rules: updates.cascadeRules,
				});
				const document = transformDocument(response);
				update(state => ({
					...state,
					currentDocument: document,
					documents: state.documents.map(d =>
						d.id === documentId ? document : d
					),
				}));
				return document;
			} catch (error: any) {
				update(state => ({ ...state, error: error.message }));
				return null;
			}
		},

		async deleteDocument(documentId: string) {
			try {
				await api.delete(`/api/documents/${documentId}`);
				update(state => ({
					...state,
					documents: state.documents.filter(d => d.id !== documentId),
					currentDocument:
						state.currentDocument?.id === documentId ? null : state.currentDocument,
					total: state.total - 1,
				}));
				return true;
			} catch (error: any) {
				update(state => ({ ...state, error: error.message }));
				return false;
			}
		},

		setCurrentDocument(document: Document | null) {
			update(state => ({ ...state, currentDocument: document }));
		},

		clearError() {
			update(state => ({ ...state, error: null }));
		},

		reset() {
			set(initialState);
		},
	};
}

export const documents = createDocumentsStore();

// Derived stores
export const currentDocument = derived(documents, $documents => $documents.currentDocument);
export const documentList = derived(documents, $documents => $documents.documents);

File: backend/routers/documents.py
"""
Document management endpoints.
"""

from datetime import datetime
from typing import Optional, List

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy import select, desc
from sqlalchemy.ext.asyncio import AsyncSession

from database import (
    get_db, User, Document, DocumentAssumption,
    DocumentGoalConnection, DocumentProgress
)
from routers.auth import get_current_user, generate_id
from utils import get_or_404, paginate, to_response, to_response_list

router = APIRouter(prefix="/documents", tags=["documents"])


class CreateDocumentRequest(BaseModel):
    title: str
    sections: dict
    conversation_id: str
    format: str = "structured_json"
    domain: Optional[str] = None
    goal_title: Optional[str] = None
    goal_id: Optional[str] = None


class UpdateDocumentRequest(BaseModel):
    title: Optional[str] = None
    sections: Optional[dict] = None
    domain: Optional[str] = None
    cells: Optional[dict] = None
    cascade_rules: Optional[dict] = None


class DocumentResponse(BaseModel):
    id: str
    user_id: str
    organization_id: str
    title: str
    sections: dict
    format: str
    domain: Optional[str]
    version: str
    is_active: bool
    conversation_id: str
    goal_title: Optional[str]
    goal_id: Optional[str]
    name: Optional[str]
    cells: Optional[dict]
    cascade_rules: Optional[dict]
    created_at: datetime
    last_updated_at: datetime


class DocumentListResponse(BaseModel):
    documents: List[DocumentResponse]
    total: int


@router.post("/", response_model=DocumentResponse)
async def create_document(
    request: CreateDocumentRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Create a new document."""
    document = Document(
        id=generate_id(),
        user_id=current_user.id,
        organization_id=current_user.organization_id,
        title=request.title,
        sections=request.sections,
        conversation_id=request.conversation_id,
        format=request.format,
        domain=request.domain,
        goal_title=request.goal_title,
        goal_id=request.goal_id,
    )
    db.add(document)
    await db.commit()
    await db.refresh(document)

    return to_response(document, DocumentResponse)


@router.get("/", response_model=DocumentListResponse)
async def list_documents(
    limit: int = Query(20, ge=1, le=100),
    offset: int = Query(0, ge=0),
    domain: Optional[str] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """List user's documents."""
    query = select(Document).where(
        Document.user_id == current_user.id,
        Document.is_active == True
    )

    if domain:
        query = query.where(Document.domain == domain)

    documents, total = await paginate(db, query, offset, limit, Document.last_updated_at)

    return DocumentListResponse(
        documents=to_response_list(documents, DocumentResponse),
        total=total,
    )


@router.get("/{document_id}", response_model=DocumentResponse)
async def get_document(
    document_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Get a specific document."""
    document = await get_or_404(db, Document, document_id, user_id=current_user.id)
    return to_response(document, DocumentResponse)


@router.patch("/{document_id}", response_model=DocumentResponse)
async def update_document(
    document_id: str,
    request: UpdateDocumentRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Update a document."""
    document = await get_or_404(db, Document, document_id, user_id=current_user.id)

    if request.title is not None:
        document.title = request.title
    if request.sections is not None:
        document.sections = request.sections
    if request.domain is not None:
        document.domain = request.domain
    if request.cells is not None:
        document.cells = request.cells
    if request.cascade_rules is not None:
        document.cascade_rules = request.cascade_rules

    document.last_updated_at = datetime.utcnow()
    await db.commit()
    await db.refresh(document)

    return to_response(document, DocumentResponse)


@router.delete("/{document_id}")
async def delete_document(
    document_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Soft delete a document."""
    document = await get_or_404(db, Document, document_id, user_id=current_user.id)

    document.is_active = False
    document.last_updated_at = datetime.utcnow()
    await db.commit()

    return {"status": "success"}
